core/drivers/crypto/caam/caam_jr.c

```diff
@@ -0,0 +1,781 @@
+
+	if ((!jr_priv->inrings) || (!jr_priv->outrings) ||
+		(!jr_priv->callers)) {
+		JR_TRACE("JR resources allocation error");
+		goto end_alloc;
+	}
+
+	/* Initialize the spin locks */
+	jr_priv->inlock = SPINLOCK_UNLOCK;
+	jr_priv->outlock = SPINLOCK_UNLOCK;
+	jr_priv->callers_lock = SPINLOCK_UNLOCK;
+
+	/* Initialize the queue counter */
+	jr_priv->inwrite_index = jr_priv->outread_index = 0;
+
+	/*
+	 * Ensure that allocated queue initialization is pushed to the physical
+	 * memory
+	 */
+	cache_operation(TEE_CACHEFLUSH, jr_priv->inrings,
+					nbJobs * sizeof(struct inring_entry));
+	cache_operation(TEE_CACHEFLUSH, jr_priv->outrings,
+					nbJobs * sizeof(struct outring_entry));
+
+	retstatus = CAAM_NO_ERROR;
+end_alloc:
+	if (retstatus != CAAM_NO_ERROR)
+		do_jr_free(jr_priv);
+	else
+		*privdata = jr_priv;
+
+	return retstatus;
+}
+
+/*
+ * Job Ring Interrupt handler
+ *
+ * Inputs:
+ * handler  Interrupt Handler structure
+ *
+ * Returns:
+ * ITRR_HANDLED  Interrupt has been handled
+ */
+static enum itr_return caam_jr_irqhandler(struct itr_handler *handler)
+{
+	struct jr_privdata *jrpriv = handler->data;
+
+	JR_TRACE("Disable the interrupt");
+	itr_disable(jrpriv->it_handler.it);
+
+	sev();
+
+	return ITRR_HANDLED;
+}
+
+
+/**
+ * Dequeues all Jobs completed. Call the job context callback
+ * function. Function returns the bit maks of the completed job
+ * expected (\a waitJobIds parameter)
+ *
+ * Inputs:
+ * waitJobIds  Expected Jobs to be complete
+ *
+ * Returns:
+ * Bit mask of the jobs completed function of input parameter
+ */
+static uint32_t do_jr_dequeue(uint32_t waitJobIds)
+{
+	uint32_t retJobId = 0;
+
+	struct caller_info   *caller = NULL;
+	struct outring_entry *jr_out = NULL;
+	struct jr_jobctx     *jobctx = NULL;
+	uint32_t  exceptions = 0;
+	int       found = -1;
+	uint16_t  idx_jr = 0;
+	uint16_t  nbJobs_done = 0;
+	size_t    nbJobs_inv = 0;
+
+	exceptions = cpu_spin_lock_xsave(&jr_privdata->inlock);
+
+	nbJobs_done = hal_jr_get_nbJobDone(jr_privdata->baseaddr);
+
+	if (nbJobs_done == 0) {
+		cpu_spin_unlock_xrestore(&jr_privdata->inlock, exceptions);
+		return retJobId;
+	}
+
+	/* Ensure that output ring descriptor entries are not in cache */
+	if ((jr_privdata->outread_index + nbJobs_done) > jr_privdata->nbJobs) {
+		/* Invalidate all job buffer */
+		/* Some job completed are at before the current outread_index */
+		jr_out     = jr_privdata->outrings;
+		nbJobs_inv = jr_privdata->nbJobs;
+	} else {
+		/* Invalidate only the job index completed */
+		jr_out = &jr_privdata->outrings[jr_privdata->outread_index];
+		nbJobs_inv = nbJobs_done;
+	}
+
+	cache_operation(TEE_CACHEINVALIDATE, (void *)jr_out,
+			(sizeof(struct outring_entry) * nbJobs_inv));
+
+	do {
+		jr_out = &jr_privdata->outrings[jr_privdata->outread_index];
+
+		/*
+		 * Lock the caller information array because enqueue is
+		 * also touching it
+		 */
+		cpu_spin_lock(&jr_privdata->callers_lock);
```

Comment:
It seems that this lock i only taken when already holding `inlock`. So this is perhaps not needed, or will it be needed later on?

Reply:
This spin lock is to protect the access to the jr_privdata->callers array that can be modified in same time when a new Job is enqueued.
This is necessary (and tested) in case of mutli-core. One core can push a new job while another fetches the completed jobs.

Reply:
In the current state of the code you're always holding `inlock` before you're taking `callers_lock`. So I'd say that it's not needed right now, unless there's a bug in the code. If it will be needed later on as the code will evolve with more support it's fine to keep it.

Reply:
Ah yes, correct my mistake I'll change to use the outlock spinlock object in the dequeue.